### ISSUE TITLE
Write a ton more Async tests

### ISSUE THREAD

**Issue by djspiewak** [typelevel/cats-effect#1046]
I fear it. Also `Deferred` seems to do better than `IOSpec` at catching things. That's not good. We should have a ton more tests which just outright-torture it, particularly with crazy cancelation cases. One random idea off hand:

Spawn tens of thousands of fibers *in parallel*, each consisting of an `async` with a finalizer. The `async` itself should call `executionContext` and use `execute` to delay its callback, but not forever. When the registration completes, it should increment an `AtomicInteger`. When the finalizer runs, it should decrement it. Once all fibers are spawn, *cancel* them all in parallel. Assert that the `AtomicInteger` is at 0.

That kind of thing.
